controllers/awsmachine_controller.go

```diff
@@ -448,57 +448,52 @@ func (r *AWSMachineReconciler) getOrCreate(scope *scope.MachineScope, ec2svc ser
+	if scope.UseSecretsManager() {
+		compressedUserData, err := userdata.GzipBytes(userData)
 		if err != nil {
-			r.Recorder.Eventf(scope.AWSMachine, corev1.EventTypeWarning, "FailedGetBootstrapData", err.Error())
 			return nil, err
 		}
-
-		if scope.UseSecretsManager() {
-			compressedUserData, err := userdata.GzipBytes(userData)
-			if err != nil {
-				return nil, err
-			}
-			// Do an initial check in case manager was terminated prematurely
-			if scope.GetSecretPrefix() == "" {
-				prefix, chunks, err := secretSvc.Create(scope, compressedUserData)
-				if err := scope.SetSecretPrefix(prefix); err != nil {
-					return nil, err
-				}
-				if err := scope.SetSecretCount(chunks); err != nil {
-					return nil, err
-				}
-				if err != nil {
-					r.Recorder.Eventf(scope.AWSMachine, corev1.EventTypeWarning, "FailedCreateAWSSecretsManagerSecrets", err.Error())
-					scope.Error(err, "Failed to create AWS Secret entry", "secretPrefix", prefix)
-					delErr := secretSvc.Delete(scope)
-					if delErr != nil {
-						scope.Error(delErr, "Failed to clean up AWS Secret entries", "secretPrefix", prefix)
-						return nil, delErr
-					}
-					return nil, err
-				}
-			}
-			// Register the Secret ARN immediately to avoid orphaning AWS resources on delete
-			if err := scope.PatchObject(); err != nil {
+		prefix, chunks, serviceErr := secretSvc.Create(scope, compressedUserData)
+		// Only persist the AWS Secrets Manager entries if at least one
+		if chunks > 0 {
+			if err := scope.SetSecretPrefix(prefix); err != nil {
 				return nil, err
 			}
-			encryptedCloudInit, err := secretsmanager.GenerateCloudInitMIMEDocument(scope.GetSecretPrefix(), scope.GetSecretCount(), scope.AWSCluster.Spec.Region)
-			if err != nil {
-				r.Recorder.Eventf(scope.AWSMachine, corev1.EventTypeWarning, "FailedGenerateAWSSecretsManagerCloudInit", err.Error())
+			if err := scope.SetSecretCount(chunks); err != nil {
 				return nil, err
 			}
-			userData = encryptedCloudInit
 		}
-
-		instance, err = ec2svc.CreateInstance(scope, userData)
+		// Register the Secret ARN immediately to avoid orphaning whatever AWS resources having been created
+		if err := scope.PatchObject(); err != nil {
+			return nil, err
+		}
+		if serviceErr != nil {
+			r.Recorder.Eventf(scope.AWSMachine, corev1.EventTypeWarning, "FailedCreateAWSSecretsManagerSecrets", serviceErr.Error())
+			scope.Error(serviceErr, "Failed to create AWS Secret entry", "secretPrefix", prefix)
+			return nil, serviceErr
+		}
+		encryptedCloudInit, err := secretsmanager.GenerateCloudInitMIMEDocument(scope.GetSecretPrefix(), scope.GetSecretCount(), scope.AWSCluster.Spec.Region)
 		if err != nil {
-			return nil, errors.Wrapf(err, "failed to create AWSMachine instance")
+			r.Recorder.Eventf(scope.AWSMachine, corev1.EventTypeWarning, "FailedGenerateAWSSecretsManagerCloudInit", err.Error())
+			return nil, err
 		}
+		userData = encryptedCloudInit
+	}
+
+	instance, err = ec2svc.CreateInstance(scope, userData)
```

Comment:
The diff is hard to read, is this inside the `if scope.UseSecretsManager() {` or outside?

Reply:
if you check the code out it's waaaay easier to read now. This is outside the block you mention.